### PR TITLE
Loosen type condition in pooling functions

### PR DESCRIPTION
**Pull request by okuta**
These fcuntions accept an array of numpy.float16, numpy.float32 and numpy.float64.
See also #555

### CODE REVIEW

File: chainer/functions/pooling/pooling_2d.py [hunk 1]
```diff
@@ -11,6 +11,11 @@
 if cuda.cudnn_enabled:
     cudnn = cuda.cudnn
     libcudnn = cudnn.cudnn
+    _cudnn_version = libcudnn.getVersion()
+
+
+def _check_cudnn_acceptable_type(x_dtype):
```
Comment (unnonouno) [inline chainer/functions/pooling/pooling_2d.py]: Does it require to check if `x_dtype.kind == 'f'`?

Comment (okuta) [inline chainer/functions/pooling/pooling_2d.py]: `x_dtype.kind` is checked at `check_type_forward`.
Do we need check again?

Comment (unnonouno) [inline chainer/functions/pooling/pooling_2d.py]: Ok, no problem.

File: tests/chainer_tests/functions_tests/pooling_tests/test_average_pooling_2d.py [hunk 2]
```diff
@@ -71,30 +80,36 @@ def test_backward_gpu_no_cudnn(self):
...
 class TestAveragePooling2DCudnnCall(unittest.TestCase):
 
     def setUp(self):
         self.x = cuda.cupy.arange(
-            2 * 3 * 4 * 3, dtype=numpy.float32).reshape(2, 3, 4, 3)
+            2 * 3 * 4 * 3, dtype=self.dtype).reshape(2, 3, 4, 3)
         self.gy = cuda.cupy.random.uniform(-1, 1,
-                                           (2, 3, 2, 2)).astype(numpy.float32)
+                                           (2, 3, 2, 2)).astype(self.dtype)
+        self.skip_test = (self.cudnn and self.dtype == numpy.float16 and
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_average_pooling_2d.py]: Does `@unittest.skip` work?

File: tests/chainer_tests/functions_tests/pooling_tests/test_unpooling_2d.py [hunk 3]
```diff
@@ -12,39 +13,50 @@
...
-    # standard output size which is estimated with get_deconv_outsize function
-    {'cover_all': False, 'outsize': (4, 2)},
-    {'cover_all': True, 'outsize': (3, 1)},
-    {'cover_all': False, 'outsize': None, 'expected_outsize': (4, 2)},
-    {'cover_all': True, 'outsize': None, 'expected_outsize': (3, 1)},
-    # another sizes which can be outsize of insize (2, 1)
-    {'cover_all': False, 'outsize': (5, 2)},
-    {'cover_all': True, 'outsize': (4, 2)},
-)
+@testing.parameterize(*[dict(x, **y) for x, y in itertools.product(
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_unpooling_2d.py]: How about making a helper function?

File: tests/chainer_tests/functions_tests/pooling_tests/test_average_pooling_2d.py (hunk 2 above)
Comment (okuta) [inline tests/chainer_tests/functions_tests/pooling_tests/test_average_pooling_2d.py]: We can't use `self` in argument of decorator.

File: tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py [hunk 4]
```diff
@@ -134,19 +122,21 @@ def test_invalid_dtype_gpu(self):
...
 
     def setUp(self):
         shape = (2, 3, 9, 8)
         size = 2 * 3 * 9 * 8
-        self.x = cuda.cupy.arange(size, dtype=numpy.float32).reshape(shape)
+        self.x = cuda.cupy.arange(size, dtype=self.dtype).reshape(shape)
         self.gy = cuda.cupy.random.uniform(
-            -1, 1, (2, 63, 1, 1)).astype(numpy.float32)
+            -1, 1, (2, 63, 1, 1)).astype(self.dtype)
+        self.skip_test = (self.cudnn and self.dtype == numpy.float16 and
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py]: Can you use `unittest.skip`?

File: tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py [hunk 5]
```diff
@@ -94,39 +99,36 @@ def test_backward_gpu_no_cudnn(self):
...
 class TestMaxPooling2DCudnnCall(unittest.TestCase):
 
     def setUp(self):
         self.x = cuda.cupy.arange(
-            2 * 3 * 4 * 3, dtype=numpy.float32).reshape(2, 3, 4, 3)
+            2 * 3 * 4 * 3, dtype=self.dtype).reshape(2, 3, 4, 3)
         self.gy = cuda.cupy.random.uniform(-1, 1,
-                                           (2, 3, 2, 2)).astype(numpy.float32)
+                                           (2, 3, 2, 2)).astype(self.dtype)
+        self.skip_test = (self.cudnn and self.dtype == numpy.float16 and
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py]: `unittest.skip`?

**Comment by unnonouno** [comment]
Use `unittest.skip` if possible.

**Comment by okuta** [comment]
@unnonouno I fixed.

**Comment by unnonouno** [comment]
Spacial pyramid pooling is unstable: https://travis-ci.org/pfnet/chainer/jobs/132084259

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**